openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -0,0 +1,149 @@
+"""
+Get call stacks of Model Class
+in three cases-
+1. QuerySet API
+2. save()
+3. delete()
+
+classes:
+CallStackManager -  stores all stacks in global dictionary and logs
+CallStackMixin - used for Model save(), and delete() method
+
+Functions:
+capture_call_stack - global function used to store call stack
+
+Decorators:
+donottrack - mainly for the places where we know the calls. This decorator will let us not to track in specified cases
+
+How to use-
+1. Import following in the file where class to be tracked resides
+    from openedx.core.lib.call_stack_manager import CallStackManager, CallStackMixin
+2. Override objects of default manager by writing following in any model class which you want to track-
+    objects = CallStackManager()
+3. For tracking Save and Delete events-
+    Use mixin called "CallStackMixin"
+    For ex.
+        class StudentModule(CallStackMixin, models.Model):
+"""
+
+import logging
+import traceback
+import re
+import collections
+from django.db.models import Manager
+
+log = logging.getLogger(__name__)
+
+# Module Level variables
+# dictionary which stores call stacks.
+# { "ModelClasses" : [ListOfFrames]}
+# Frames - ('FilePath','LineNumber','Context')
+# ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
+#                                                     [(file,line number,context),(---,---,---)]]}
+STACK_BOOK = collections.defaultdict(list)
+
+# filter to trickle down call stacks
+EXCLUDE = ['^.*python2.7.*$', '^.*<exec_function>.*$', '^.*exec_code_object.*$']
+REGULAR_EXPS = [re.compile(x) for x in EXCLUDE]
+
+# Variable which decides whether to track calls in the function or not. Do it by default.
+TRACK_FLAG = True
+
+# List keeping track of Model classes not be tracked for special cases
+# usually cases where we know that the function is calling Model classes.
+HALT_TRACKING = []
+
+
+def capture_call_stack(current_model):
+    """ logs customised call stacks in global dictionary `STACK_BOOK`, and logs it.
+
+    Args:
+        current_model - Name of the model class
+    """
+    # holds temporary callstack
+    # frame[0][6:-1] -> File name along with path
+    # frame[1][6:] -> Line Number
+    # frame[2][3:] -> Context
+    temp_call_stack = [(frame[0][6:-1],
+                        frame[1][6:],
+                        frame[2][3:])
+                       for frame in [stack.replace("\n", "").strip().split(',') for stack in traceback.format_stack()]
+                       if not any(reg.match(frame[0]) for reg in REGULAR_EXPS)]
+
+    # avoid duplication.
+    if temp_call_stack not in STACK_BOOK[current_model] and TRACK_FLAG \
+            and not any(issubclass(current_model, cls) for cls in HALT_TRACKING):
```

Comment:
I think it would be clearer to use the `tuple` syntax you suggested above, rather than using `any` and multiple calls to `issubclass`.